plugins/protocols/session-lock.cpp

```diff
@@ -0,0 +1,434 @@
+                    wo->width / wo->scale,
+                    wo->height / wo->scale);
+                LOGD("surface_configure output=", wo->name, " ", wo->width, "x", wo->height);
+
+                auto output = wf::get_core().output_layout->find_output(lock_surface->output);
+                if (!output || (output_states.find(output) == output_states.end()))
+                {
+                    LOGE("lock_surface created on deleted output ", wo->name);
+                    return;
+                }
+
+                // TODO: hook into output size changes and reconfigure.
+
+                output_states[output]->surface = std::make_shared<lock_surface_node_t>(
+                    this, lock_surface, output);
+
+                if (state == LOCKED)
+                {
+                    // Output is already inhibited.
+                    output_states[output]->surface->attach_to_layer();
+                } else if (have_all_surfaces())
+                {
+                    // All lock surfaces ready. Lock.
+                    lock_timer.disconnect();
+                    lock_all();
+                }
+            });
+            new_surface.connect(&lock->events.new_surface);
+
+            unlock.set_callback([this] (void *data)
+            {
+                unlock_all();
+            });
+            unlock.connect(&lock->events.unlock);
+
+            destroy.set_callback([this] (void *data)
+            {
+                output_added.disconnect();
+                output_removed.disconnect();
+                new_surface.disconnect();
+                unlock.disconnect();
+                destroy.disconnect();
+                set_state(state == UNLOCKED ? DESTROYED : ZOMBIE);
+                LOGD("session lock destroyed");
+            });
+            destroy.connect(&lock->events.destroy);
+
+            lock_timer.set_timeout(1000, [this] (void)
+            {
+                lock_all();
+            });
+        }
+
+        void surface_destroyed(wf::output_t *output)
+        {
+            if (output_states.find(output) != output_states.end())
+            {
+                output_states[output]->surface.reset();
+                if (output_states[output]->backup_surface)
+                {
+                    output_states[output]->backup_surface->display();
+                }
+            }
+        }
+
+        ~wayfire_session_lock()
+        {
+            remove_backup_surfaces();
+        }
+
+      private:
+        void handle_output_added(wf::output_t *output)
+        {
+            output_states[output] = std::make_shared<output_state>(output);
+            if (state == LOCKED)
+            {
+                lock_output(output, output_states[output]);
+            }
+        }
+
+        void handle_output_removed(wf::output_t *output)
+        {
+            output_states.erase(output);
+            auto next_output_state = output_states[wf::get_core().seat->get_active_output()];
```

Comment:
Again, is this needed? Core should auto-refocus correctly if you filter by the output in the refocus_keyboard() method.